### PR TITLE
feat: Add a check that validates Gtk 3 installation on Windows

### LINKED ISSUE #12: Add a check that validates Gtk 3 installation on Windows
When running `Uno-check`, I am currently getting a message **Congratulations, everything looks great!**. However, I am unable to run the Skia.Gtk template application because I am missing the gtk runtimes. It would be awesome for the tool to detect that.  To make sure the user has the version of Gtk, and other requirements, that are required by the default templates.

### PR DESCRIPTION
**Pull request by aesteves900**
GitHub Issue (If applicable): closes #12  

# PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

People from the community have reported that certain projects fail to build because of long paths.

There is no check for GTK3 activation on Windows.

## What is the new behavior?

The uno-check will validate the presence of GTK3+ and install if required.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current Figma desktop app
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] Unit Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes and the __PlugIn is BACKWARD COMPATIBLE__
- [x] Associated with an issue

### CODE REVIEW

File: UnoCheck/Checkups/WindowsPhytonInstallationCheckup.cs [hunk 1]
```diff
@@ -17,25 +16,35 @@ public class WindowsPhytonInstallationCheckup : Checkup
 
 		public override async Task<DiagnosticResult> Examine(SharedState history)
 		{
-			RegistryKey pyLaucherKey = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Python\PyLauncher", false);
-			string? path = pyLaucherKey?.GetValue(InstallDirKey)?.ToString();
-
-			if (pyLaucherKey is null || string.IsNullOrEmpty(path))
+			if (!this.PhytonIsPresent())
```
Comment (jeromelaban) [inline UnoCheck/Checkups/WindowsPhytonInstallationCheckup.cs:19]: This looks unrelated ?

File: UnoCheck/Process/ShellProcessRunner.cs [hunk 2]
```diff
@@ -23,11 +21,11 @@ public ShellProcessRunnerOptions(string exe, string args, CancellationToken canc
 			Args = args;
 			CancellationToken = cancellationToken;
 		}
-		 
+
 		public string Executable { get; set; }
 		public string Args { get; set; }
 
-		public bool Verbose { get;set; }
+		public bool Verbose { get; set; }
```
Comment (jeromelaban) [inline UnoCheck/Process/ShellProcessRunner.cs:28]: Those changes look unrelated, make sure to remove them before removing the draft status.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace DotNetCheck.Solutions
+{
+	public class GTK3NotInstalledOnWindowsSolution : Solution
+	{
+		public override async Task Implement(SharedState sharedState, CancellationToken cancellationToken)
+		{
+			StringBuilder stringBuilder = new();
+
+			var chocoResult = ShellProcessRunner.Run("powershell.exe", "Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))");
```
Comment (jeromelaban) [inline UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:14]: We can suggest to install a specific version, or install one particular binary "https://github.com/tschoonj/GTK-for-Windows-Runtime-Environment-Installer/releases". There's no need to install everything else.

File: UnoCheck/Checkups/WindowsPhytonInstallationCheckup.cs (hunk 1 above)
Comment (aesteves900) [reply UnoCheck/Checkups/WindowsPhytonInstallationCheckup.cs:19]: I found an issue when I working on GTK3. I just changed the way to detect Python.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs (hunk 3 above)
Comment (aesteves900) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs]: Done!

File: UnoCheck/Process/ShellProcessRunner.cs (hunk 2 above)
Comment (aesteves900) [reply UnoCheck/Process/ShellProcessRunner.cs:30]: Done!

File: UnoCheck/RuntimePlatform/Core/IRuntimePlatform.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+{
+	public interface IRuntimePlatform
+	{
+		Platform Platform { get; }
+
+		IList<InstalledProgramInfo> GetInstalledPrograms();
+
+		bool IsInstalledProgram(string partialName);
+	}
+}
```
Comment (jeromelaban) [inline UnoCheck/RuntimePlatform/Core/IRuntimePlatform.cs:14]: I do not see the point of this interface (and the class RuntimePlatformBase) at this time. Let's remove it for now and we'll add it when we'll need to make generic code based on that.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace DotNetCheck.Solutions
+{
+	public class GTK3NotInstalledOnWindowsSolution : Solution
+	{
+		public override async Task Implement(SharedState sharedState, CancellationToken cancellationToken)
+		{
+			byte[] data = null;
+
+			WebClient client = new();
```
Comment (jeromelaban) [inline UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:16]: There's already a solution that exists which does something very similar. Reuse or adjust he code from:

https://github.com/unoplatform/uno.check/blob/1aa1a2c0df19c860b71c68b0d3b356cc21800e15/UnoCheck/Solutions/MsInstallerSolution.cs#L42

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+			else
+			{
+				ReportStatus($"GTK3+ is {nameof(Status.Error)}");
+				throw new InvalidOperationException("GTK3+ cannot be installed.");
+			}
+
+			await Task.CompletedTask;
+		}
+
+		private const string GtkUrl = "https://github.com/tschoonj/GTK-for-Windows-Runtime-Environment-Installer/releases/download/2022-01-04/gtk3-runtime-3.24.31-2022-01-04-ts-win64.exe";
```
Comment (jeromelaban) [inline UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:50]: This should come from the manifest file: 

https://github.com/unoplatform/uno.check/blob/1aa1a2c0df19c860b71c68b0d3b356cc21800e15/manifests/uno.ui.manifest.json

File: UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+			var registry_key = @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall";
+
+			using (RegistryKey key = RegistryKey.OpenRemoteBaseKey(RegistryHive.LocalMachine, machine).OpenSubKey(registry_key))
+			{
+				foreach (string subkey_name in key.GetSubKeyNames())
+				{
+					using (RegistryKey subkey = key.OpenSubKey(subkey_name))
+					{
+						if (subkey.GetValue("DisplayName") != null)
```
Comment (jeromelaban) [inline UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs:33]: I wonder, DisplayName may be localized. Isn't there an "ID" or Guid we can use instead?

File: UnoCheck/ToolInfo.cs [hunk 8]
```diff
@@ -43,6 +41,12 @@ public static string CurrentVersion
 
 		public static bool Validate(Manifest.Manifest manifest)
 		{
+			/* Unmerged change from project 'UnoCheck (net5.0)'
+			Before:
+						DotNetCheck.WindowsPlatform.Windows.CreateInstalledProgramsInventory();
+			After:
+						Windows.CreateInstalledProgramsInventory();
+			*/
```
Comment (jeromelaban) [inline UnoCheck/ToolInfo.cs:49]: What's the intent of that comment?

File: UnoCheck/Manifest/Gtk.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+		[JsonProperty("minimumVersion")]
+		public string MinimumVersion { get; set; }
+
+		[JsonIgnore]
+		public string CompatVersion
+			=> Version ?? MinimumVersion;
+
+		[JsonProperty("requireExact")]
+		public bool RequireExact { get; set; }
```
Comment (jeromelaban) [inline UnoCheck/Manifest/Gtk.cs:25]: Are these needed at this point? (Maybe by the `Get` method?)

File: UnoCheck/RuntimePlatform/Core/IRuntimePlatform.cs (hunk 4 above)
Comment (aesteves900) [reply UnoCheck/RuntimePlatform/Core/IRuntimePlatform.cs]: I removed!

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs (hunk 6 above)
Comment (aesteves900) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs]: Done!

File: UnoCheck/ToolInfo.cs (hunk 8 above)
Comment (aesteves900) [reply UnoCheck/ToolInfo.cs]: It was removed!

File: UnoCheck/Manifest/Gtk.cs (hunk 9 above)
Comment (aesteves900) [reply UnoCheck/Manifest/Gtk.cs]: I removed the RequireExtac because is not used. I don't know if I got your point, please let me know any change.

File: UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs (hunk 7 above)
Comment (aesteves900) [reply UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs:33]: GTK is an acronym, no worries with localization.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs (hunk 5 above)
Comment (aesteves900) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs]: Done!

File: UnoCheck/Manifest/Gtk.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace DotNetCheck.Manifest
+{
+	public partial class Gtk
+	{
+		[JsonProperty("urls")]
+		public Urls Urls { get; set; }
+
+		[JsonIgnore]
+		public System.Uri Url
```
Comment (ebariche) [inline UnoCheck/Manifest/Gtk.cs:11]: Remove. Not used.

File: UnoCheck/Manifest/Gtk.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+			=> Urls?.Get(CompatVersion);
+
+		[JsonProperty("version")]
+		public string Version { get; set; }
+
+		[JsonProperty("minimumVersion")]
+		public string MinimumVersion { get; set; }
```
Comment (ebariche) [inline UnoCheck/Manifest/Gtk.cs:18]: Remove. Not used.

File: UnoCheck/Manifest/Gtk.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+		[JsonIgnore]
+		public string CompatVersion
```
Comment (ebariche) [inline UnoCheck/Manifest/Gtk.cs:21]: Remove. Not used.

File: UnoCheck/Manifest/Gtk.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
+﻿using Newtonsoft.Json;
+
+namespace DotNetCheck.Manifest
+{
+	public partial class Gtk
```
Comment (ebariche) [inline UnoCheck/Manifest/Gtk.cs:5]: partial ?

File: UnoCheck/ToolInfo.cs [hunk 14]
```diff
@@ -1,10 +1,8 @@
 ﻿using NuGet.Versioning;
 using Spectre.Console;
 using System;
-using System.Collections.Generic;
```
Comment (ebariche) [inline UnoCheck/ToolInfo.cs:4]: Undo these changes, this is not related to this PR.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+		{
+			Url = url;
+			Title = title;
+		}
+
+		public Uri Url { get; set; }
+		public string Title { get; set; }
+
+		public override async Task Implement(SharedState sharedState, CancellationToken cancelToken)
+		{
```
Comment (ebariche) [inline UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:23]: Since we're duplicating this code, shouldn't this be refactored ? @jeromelaban

File: UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
+﻿using DotNetCheck.RuntimePlatform.Core;
+using DotNetCheck.RuntimePlatform.Model;
+using Microsoft.Win32;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DotNetCheck.RuntimePlatform.OperationalSystem
+{
+	public class WindowsRuntime : RuntimePlatformBase
```
Comment (ebariche) [inline UnoCheck/RuntimePlatform/OperationalSystem/WindowsRuntime.cs:10]: I would use a different name here. "WindowsRuntime" is a Microsoft technology and is not an OS.

File: UnoCheck/Manifest/Gtk.cs (hunk 13 above)
Comment (aesteves900) [reply UnoCheck/Manifest/Gtk.cs:5]: All classes are marked as partial. Can I remove it anyway?

image

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs (hunk 15 above)
Comment (aesteves900) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:23]: Hello @jeromelaban, what do you want to do about that?

File: UnoCheck/Manifest/Gtk.cs (hunk 13 above)
Comment (jeromelaban) [reply UnoCheck/Manifest/Gtk.cs:5]: Let's keep it partial for now, it does not change much.

File: UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs (hunk 15 above)
Comment (jeromelaban) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:23]: Let's open an issue to take a look independently.

Comment (aesteves900) [reply UnoCheck/Solutions/GTK3NotInstalledOnWindowsSolution.cs:23]: I was created the issue #96

File: UnoCheck/Checkups/VisualStudioWindowsCheckup.cs [hunk 17]
```diff
@@ -46,7 +46,7 @@ public override async Task<DiagnosticResult> Examine(SharedState history)
 			{
 				ReportStatus($"Missing Visual Studio >= {MinimumVersion.ThisOrExact(ExactVersion)}", Status.Error);
 
-				return new DiagnosticResult(Status.Error, this);
+				return new DiagnosticResult(Status.Warning, this);
```
[the remark this answers is not in the document]
Comment (jeromelaban) [reply UnoCheck/Checkups/VisualStudioWindowsCheckup.cs:49]: @aesteves900

File: UnoCheck/Manifest/Gtk.cs (hunk 10 above)
Comment (jeromelaban) [reply UnoCheck/Manifest/Gtk.cs:11]: @aesteves900

**Comment by jeromelaban** [comment]
Deprecated feature.
